message/messaging_service.cc

```diff
@@ -606,18 +606,26 @@ gms::inet_address messaging_service::get_preferred_ip(gms::inet_address ep) {
 
 void messaging_service::init_local_preferred_ip_cache(const std::unordered_map<gms::inet_address, gms::inet_address>& ips_cache) {
     _preferred_ip_cache = ips_cache;
+    _preferred_to_endpoint.clear();
     //
     // Reset the connections to the endpoints that have entries in
     // _preferred_ip_cache so that they reopen with the preferred IPs we've
     // just read.
     //
     for (auto& p : _preferred_ip_cache) {
         this->remove_rpc_client(msg_addr(p.first));
+        _preferred_to_endpoint[p.second] = p.first;
     }
 }
 
 void messaging_service::cache_preferred_ip(gms::inet_address ep, gms::inet_address ip) {
     _preferred_ip_cache[ep] = ip;
+    _preferred_to_endpoint[ip] = ep;
+}
+
+gms::inet_address messaging_service::get_public_endpoint_for(const gms::inet_address& ip) const {
+    auto i = _preferred_to_endpoint.find(ip);
+    return i != _preferred_to_endpoint.end() ? i->second : ip;
```

Comment:
The map is public ip  to internal ip. So  i->second gives a internal ip, no? However the name get_public_endpoint_for suggests this function will return a public ip not an internal ip.

Reply:
No, `_preferred_ip_cache` is public to private. This is a reverse mapping, private to public. So the name suggests exactly what it does.

Reply:
So, the comment here is wrong for the new map _preferred_to_endpoint. This is confusing. 

```
    // map: Node broadcast address -> Node internal IP for communication within the same data center
    std::unordered_map<gms::inet_address, gms::inet_address> _preferred_ip_cache, _preferred_to_endpoint;
``

Reply:
OK. No problem.